controller/konnect/reconciler_kongplugin.go

```diff
@@ -88,130 +103,124 @@ func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request)
-			for _, pb := range bindingMapping.byServiceName[kongService.Name] {
-				if lo.ContainsBy(pb.OwnerReferences, func(ownerRef metav1.OwnerReference) bool {
-					if ownerRef.Kind == "KongPlugin" && ownerRef.Name == kongPlugin.Name && ownerRef.UID == kongPlugin.UID {
-						return true
-					}
-					return false
-				}) {
-					// The PluginBinding is dangling, so it needs to be deleted
-					pluginBindingsToDelete = append(pluginBindingsToDelete, pb)
-				} else {
-					pluginReferenceFound = true
-				}
+	for cpNN, relations := range groupedCombinations {
+		for _, rel := range relations {
+			pluginReferenceFound = true
+
+			builder := NewKongPluginBindingBuilder().
+				WithGenerateName(kongPlugin.Name + "-").
+				WithNamespace(kongPlugin.Namespace).
+				WithPluginRef(kongPlugin.Name).
+				WithControlPlaneRef(&configurationv1alpha1.ControlPlaneRef{
+					Type: configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef,
+					KonnectNamespacedRef: &configurationv1alpha1.KonnectNamespacedRef{
+						Namespace: cpNN.Namespace,
+						Name:      cpNN.Name,
+					},
+				})
+
+			l := kongPluginBindingList.Items
+
+			if rel.Service != "" {
+				l = lo.Filter(l, func(pb configurationv1alpha1.KongPluginBinding, _ int) bool {
+					return pb.Spec.Targets.ServiceReference != nil &&
+						pb.Spec.Targets.ServiceReference.Name == rel.Service
+				})
+				builder.WithServiceTarget(rel.Service)
 			}
-		} else {
-			pluginSlice = strings.Split(plugins, ",")
-
-			for _, pb := range kongPluginBindingList.Items {
-				// if the kongPluginBinding targets the KongService,
-				if pb.Spec.Targets.ServiceReference != nil &&
-					pb.Spec.Targets.ServiceReference.Name == kongService.Name &&
-					// but the service does not contain the plugin referenced by the binding in the annotation, and
-					!lo.Contains(pluginSlice, pb.Spec.PluginReference.Name) &&
-					// the plugin is managed (created out of an annotation)
-					lo.ContainsBy(pb.OwnerReferences, func(ownerRef metav1.OwnerReference) bool {
-						if ownerRef.Kind == "KongPlugin" && ownerRef.Name == kongPlugin.Name && ownerRef.UID == kongPlugin.UID {
-							return true
-						}
-						return false
-					}) {
-					// then mark it for deletion, as the plugin is not referenced by the KongService anymore
-					pluginBindingsToDelete = append(pluginBindingsToDelete, pb)
-				}
+			if rel.Route != "" {
+				l = lo.Filter(l, func(pb configurationv1alpha1.KongPluginBinding, _ int) bool {
+					return pb.Spec.Targets.RouteReference != nil &&
+						pb.Spec.Targets.RouteReference.Name == rel.Route
+				})
+				builder.WithRouteTarget(rel.Route)
+			}
+
+			// TODO consumers and consumer groups
+
+			kongPluginBinding := builder.Build()
+			if err := controllerutil.SetOwnerReference(&kongPlugin, kongPluginBinding, clientWithNamespace.Scheme(), controllerutil.WithBlockOwnerDeletion(true)); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to set owner reference: %w", err)
```

Comment:
I think this, as a part of building the object, could be done by the builder with `WithOwnerReference(kongPluginBinding)`. It will require passing a client to `NewKongPluginBindingBuilder`, but that might be required for validating the references anyway.

Reply:
c19d953d444b6b220d794613c0b7ba01be80dbc3